calibrated timestamps sample

gpx1000 opened the pull request:
## Description

Add a sample showing how to use Calibrated Timestamps in an existing sample.

Fixes #<issue number>

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

wflohry committed:
Using basic raytracing as template

wflohry committed:
Load sponza

wflohry committed:
Loading sponza model

wflohry committed:
Separating out ray tracing command buffers

wflohry committed:
Add render modes

wflohry committed:
Edit vertex structure

wflohry committed:
Put static objects in buffer with device-exclusive memory

wflohry committed:
Adding vertex and index data to bindings

wflohry committed:
Demonstration of referencing model data from ray tracing shader

wflohry committed:
Implemented texture array and visualization

wflohry committed:
Added fire pit

wflohry committed:
Added ray-traced shadows

wflohry committed:
Adding support for AO

wflohry committed:
Updated model settings

gpx1000 committed:
Fixed a few warnings and spelling errors.

Also fixed the fences problem that causes crash bug from fences on close.

wflohry committed:
Separating out static model loading

wflohry committed:
Separate handling of static / dynamic buffers

wflohry committed:
Added dynamic object and refactored

gpx1000 committed:
Fixed another warning and fixed the fences.

gpx1000 committed:
Merge branch 'feature/raytrace-extended' into warn_removal_and_fence_fix

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp

gpx1000 committed:
Switch to using constants and specializations in order to change values.  Can switch back to uniforms at a later date but the uniforms were proving a render issue for linux.  This should also demonstrate how to use multiple specialization constants in a ray traced environment.

gpx1000 committed:
more warnings removal

wflohry committed:
Merge warn_remove_and_fence_fix

gpx1000 committed:
Merged in warn_removal_and_fence_fix (pull request #1)

Fixed a few warnings and spelling errors.

Approved-by: William Lohry

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

wflohry committed:
Moving instance buffer creation to TLAS construction

wflohry committed:
Adding refraction placeholder

wflohry committed:
Perform BLAS update for dynamic objects and TLAS update every frame.

wflohry committed:
Adding dynamic sinusoidal wave

wflohry committed:
Finished refractive wave

wflohry committed:
Update to refraction

wflohry committed:
Adding separate flame instance

wflohry committed:
Working flame / particle generator

wflohry committed:
Updates to flame and refraction calculation

wflohry committed:
Updated refraction algorithm

wflohry committed:
Added texture

gpx1000 committed:
more warnings removal
move the png to ktx format

gpx1000 committed:
ensure we're pointing to the right place

wflohry committed:
Merge and update

wflohry committed:
Update to ktx commit

gpx1000 committed:
run clang-format over the sample.

wflohry committed:
Added transformations

wflohry committed:
Merged

wflohry committed:
Merged

gpx1000 committed:
Merged in clang-format (pull request #3)

run clang-format over the sample.

Approved-by: William Lohry

wflohry committed:
Adding transformations

wflohry committed:
Implemented automatic  billboard rotation

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

gpx1000 committed:
run a profile pass.
move pts, uvs, normals, indices to be created/initialized once to avoid dynamic memory generation.
remove warnings.
Update copyright.

gpx1000 committed:
Merged in profile (pull request #4)

run a profile pass.

Approved-by: William Lohry

gpx1000 committed:
Update copyright in CMakeLists.txt
Update sample cmake function definition
Update samples list in Samples CMakeLists.txt
Update README.md with first pass definition of the sample.

gpx1000 committed:
Merged in draft_ready (pull request #5)

Update copyright in CMakeLists.txt

Approved-by: William Lohry

wflohry committed:
Added tutorial rough draft

gpx1000 committed:
grammar check.

gpx1000 committed:
grammar check.

gpx1000 committed:
Merged in tutorial (pull request #6)

grammar check.

Approved-by: William Lohry

wflohry committed:
Added fence reset

wflohry committed:
Updating triangle indexing code

wflohry committed:
Avoid dynamic allocation / copy

gpx1000 committed:
update copyright

gpx1000 committed:
Merged in feature/raytrace-extended (pull request #7)

Feature/raytrace extended

gpx1000 committed:
Merge branch 'downstream-master'

gpx1000 committed:
Fixes for Clang-format and copyright headers CI failures.

gpx1000 committed:
add in nonuniformEXT to the texture lookup as imageOffset isn't uniform.

gpx1000 committed:
switch to textureLod from texture to ensure it's ready for mipmaps if copied from sample.

wflohry committed:
Fixing refraction bug

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

gpx1000 committed:
fix typo in README.md

gpx1000 committed:
fix typo in README.md and in CMakeLists.txt

gpx1000 committed:
Merge branch 'master' into master

wflohry committed:
Updates for PR comments

wflohry committed:
Adding memory barriers

wflohry committed:
Updating shaders to use defines

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp
#	shaders/khr_ray_tracing_extended/closesthit.rchit

gpx1000 committed:
clang format cleanup

gpx1000 committed:
add missing copyright to new header

gpx1000 committed:
fix attribution in the tutorial.md

gpx1000 committed:
framework:
acceleration structure: (build) fix a bug which only would allow a single triangle to be included in a acceleration structure.
acceleration structure: (build) allow for more than one call to build (useful in animation).  This was caused by recreating the buffer each time rather than reusing it if it already exists.
acceleration structure: (add triangles) allow for developer to specify the data address of the vertex index and transform buffers that are used.  This allows for multi model gltf files to work as in the sponza scene.
acceleration structure: enable buffer retrieval to allow for developer to add in memory barriers.
acceleration structure: provide developer means to wipe the geometries added to the acceleration structure.  This is very important if animating and thus recalling the same structure(s).  NB: a future feature would be to index the geometries and instances to allow for updating existing data individually rather than a requirement to wipe out everything and rebuild everything.
scratch buffer: add a return for the scratch buffer's size to allow for reusing the scratch buffer if it's the same size as previously allocated.

Sample:
As this is the first ray tracing sample to make use of @Sascha William's excellent acceleration structure framework, and we already have the existing long form, I opted to enable a macro to turn on/off using the framework's acceleration structure.  I think this is valuable to end users to show a precise "what the frameork does" behind the scenes.  Future samples should opt to use the framework.

gpx1000 committed:
run clang-format over scratch_buffer.h

gpx1000 committed:
Add in proper update logic for the acceleration_structure in both the framework and the sample.

gpx1000 committed:
Ensure that only updated geometry are used in each update pass and make framework the default AS.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Small changes for buffer.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
run clang-format

gpx1000 committed:
fix android

gpx1000 committed:
fix android

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
address the returns doc.

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge remote-tracking branch 'origin/master'

gpx1000 committed:
Fix issue #452

gpx1000 committed:
Merge branch 'upstream-master'

JeremyG2101 committed:
PLEASE DO NOT MERGE IT! This is a Debug sample including: memory_budget

JeremyG2101 committed:
working version update

JeremyG2101 committed:
Quick fix just removed 2 test log lines

JeremyG2101 committed:
All changes were made based on Steve's comments. Syntax warnings were claned from my end. All printf related functions/commands were removed.

JeremyG2101 committed:
copyrights year should be 2019 based on the instancing created

JeremyG2101 committed:
CMake list update.

JeremyG2101 committed:
add copyrights to CMake List

JeremyG2101 committed:
Per Steve's Comments: 1) In the CMakeList, the introduction is now 100% focus on what was introduced related to the memory budget extension feature. 2)update_memory_properties() function is now REMOVED from the draw() and reintroduced to the prepare_instance_data() function, after everything is calculated, loaded, and staged. 3) All runtime update/button-press-update related boolean and functions were REMOVED.

JeremyG2101 committed:
The Tutorial is fianlly done, and the image is saved in .PNG format, organized in an individual folder since the memory_budget folder. the tutorial is named as README.md

JeremyG2101 committed:
This update regards to Steve's newest comments to revise the description modification copyrights to Holochip Corporation; and insert additional description of the usage of update_device_memory_properties().

JeremyG2101 committed:
Quick syntax fix on "the" to "The" from description of int variable "mesh_density".

JeremyG2101 committed:
went thfough the clang check and already have the format modified, our project name is added to the CmakeList

JeremyG2101 committed:
Merge pull request #8 in holochip/vulkan-samples from Memory_Budget_Extension to master

Memory_Budget_Debug_Sample

Approved by Steven Winston

JeremyG2101 committed:
FR-406: Memory Budget: Quick fix corresponding for the GitHub comments regarding memory units conversion issues and get_memory_properties2 call inside prepare()

JeremyG2101 committed:
Code revised per Steve's new comments 10/25/2022

JeremyG2101 committed:
This is the dummy renderer class built for full_screen_exclusive extension sample, after it passed check, I will create an individual branch just for a dummy renderer, it I don't have to write 1,000+ lines to begin with. Where I can directly dive into add models and change skybox.

JeremyG2101 committed:
So this is the clean and neat version of it, Clang check applied.

JeremyG2101 committed:
code revised and set private to extension related variables

JeremyG2101 committed:
Full Screen Extension exclusive

JeremyG2101 committed:
Organized the full Screen Exclusive branch, which it is actually corresponding to the current/up-to-date changes of the memory_budget. It also adds the full_screen_exclusive to the project CMake list.

JeremyG2101 committed:
Adds more annotations for display mode selection

JeremyG2101 committed:
noticed an incorrect annotation

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

JeremyG2101 committed:
A clean version of this

gpx1000 committed:
Merged master into Full_Screen_Exclusive_Extension

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge remote-tracking branch 'origin/master'

gpx1000 committed:
Merge branch 'master' into downstream-master

# Conflicts:
#	samples/CMakeLists.txt

gpx1000 committed:
Revert memory budget should NOT be merged to master.

gpx1000 committed:
Revert "This update regards to Steve's newest comments to revise the description modification copyrights to Holochip Corporation; and insert additional description of the usage of update_device_memory_properties()."

This reverts commit 889ece87b111e808e71a4d98069fe7df1ccc0b68.

gpx1000 committed:
Revert "Quick syntax fix on "the" to "The" from description of int variable "mesh_density"."

This reverts commit 35e7dae9ff8a859a1905867236dd88fc6cd22fd2.

gpx1000 committed:
Revert "The Tutorial is fianlly done, and the image is saved in .PNG format, organized in an individual folder since the memory_budget folder. the tutorial is named as README.md"

This reverts commit 3cc5023045cf19ee3ad1e011f35ae84b394d9597.

gpx1000 committed:
Revert "Per Steve's Comments: 1) In the CMakeList, the introduction is now 100% focus on what was introduced related to the memory budget extension feature. 2)update_memory_properties() function is now REMOVED from the draw() and reintroduced to the prepare_instance_data() function, after everything is calculated, loaded, and staged. 3) All runtime update/button-press-update related boolean and functions were REMOVED."

This reverts commit 8cdc982c22e040ace07e62a866758e3e58e99a81.

gpx1000 committed:
Revert "copyrights year should be 2019 based on the instancing created"

This reverts commit 428cc6eab40f504a1b846995f1e9274051daae62.

gpx1000 committed:
Revert "add copyrights to CMake List"

This reverts commit 1086cab586b2dfa78eac2b111c5ff4d39779fec9.

gpx1000 committed:
Revert "CMake list update."

This reverts commit 8c14c62074c1bb713e35da3f546a6a5009de9a90.

gpx1000 committed:
Revert "All changes were made based on Steve's comments. Syntax warnings were claned from my end. All printf related functions/commands were removed."

This reverts commit 8970f8fdce9eaa76d743748982be4d06e6a5be61.

gpx1000 committed:
Revert "Quick fix just removed 2 test log lines"

This reverts commit 7708fff86dd2adf80aa37efb9e122b117c18cd7d.

gpx1000 committed:
Revert "working version update"

This reverts commit 9a56f3b7bb387f683d6b08bb89becb1110e7256d.

gpx1000 committed:
Revert "PLEASE DO NOT MERGE IT! This is a Debug sample including: memory_budget"

This reverts commit f6d3730ebcfe9a4f96451133e18fddf8b3af03ef.

gpx1000 committed:
Merge branch 'downstream-master' into Full_Screen_Exclusive_Extension

JeremyG2101 committed:
added all fixes according to the last commonts. This is also a quick save point before I add the customized swapchain, swapchain image, and image view functions. The mentioned function set will ultimately replaces the swapchain and image views created from the render_context class, but remains everything else the same.

JeremyG2101 committed:
full screen exclusive are updated based on framework changes made in the most up-to-date vulkan sample version. In addition to which, 3 bugs are fixed (suggested to fix) in response to the newest version of glslang, regarding to the directory/directories change(s) made to LimitedResources.

JeremyG2101 committed:
This update adds an entire image view recreate function in addition to the swapchain recreation. Meantime, the swapchain images are are deep copied to a new image and sync them after the swapchain recreation. Clang checked for both files.

JeremyG2101 committed:
Add a boolean to activate or deactive swapchain and full screen exclusive related functions

gpx1000 committed:
add ninja-build to the Dockerfile.

JeremyG2101 committed:
Calibrated Time Stamps

JeremyG2101 committed:
Introduces update_time_domain, read_time_domain, and update_timestamps. Several vectors and structures are introduced as well. All syntax organized in Clang format.

JeremyG2101 committed:
Quick Save: add some veriations to the subpass, added and modified calibrated timestamps related functions, device extension debug, sample CMake list updated

JeremyG2101 committed:
quick save adding some modifications

gpx1000 committed:
Merge remote-tracking branch 'upstream/main' into upstream-main

JeremyG2101 committed:
Quick Save: this pretty much finalizes the entire sample, its tutorial will be updated shortly.

JeremyG2101 committed:
Adds README.md file, and its reference image; minior formating issues and Clang format checked.

JeremyG2101 committed:
fixed some wording issues from README.md

JeremyG2101 committed:
restores framework back to master

JeremyG2101 committed:
wording issues fixed

JeremyG2101 committed:
fixes a minor issue considering time_domain_count could possibly drop down to 1, during optimal time domain update.

JeremyG2101 committed:
some wording issues fixed, cpp code reference is now up-to-date.

JeremyG2101 committed:
Fixed some outdated bllean variables which are not being used

JeremyG2101 committed:
.cpp reference is up-to-date now

gpx1000 committed:
use the same submodules as main.

gpx1000 committed:
Merge branch 'upstream-main' into vk_Calibrated_Timestamps_Sample

# Conflicts:
#	samples/CMakeLists.txt

gpx1000 committed:
didn't mean to put this as part of the PR.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+class CalibratedTimestamps : public ApiVulkanSample
+{
+  private:
+	bool                         isTimeDomainUpdated = false;        // this is just to tell if time domain update has a VK_SUCCESS in the end
```
I think, the naming convention for member variables is camel_case here. That is, isTimeDomainUpdated should probably be named is_time_domain_updated.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+{
+  private:
+	bool                         isTimeDomainUpdated = false;        // this is just to tell if time domain update has a VK_SUCCESS in the end
+	bool                         isTimestampUpdated  = false;        // this is just to tell the GUI whether timestamps operation has a VK_SUCCESS in the end
```
isTimestampUpdated -> is_timestamp_updated

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+  private:
+	bool                         isTimeDomainUpdated = false;        // this is just to tell if time domain update has a VK_SUCCESS in the end
+	bool                         isTimestampUpdated  = false;        // this is just to tell the GUI whether timestamps operation has a VK_SUCCESS in the end
+	uint32_t                     time_domain_count   = 0;
```
Why do you need `time_domain_count`? It's just the `time_domains.size()`.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+
+	struct FrameBufferAttachment
+	{
+		VkImage        image{};
```
Why do you initialize some handles with `= VK_NULL_HANDLE`, and others with `{}`?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+	std::vector<std::string> object_names{};
+
+  private:
+	// MODs:
```
`MODs:` ?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+	static std::string read_time_domain(VkTimeDomainEXT inputTimeDomain);        // this returns a human-readable info for what time domain corresponds to what
+	void               update_time_domains();                                    // this extracts total number of time domain the (physical device has, and then sync the time domain EXT data to its vector
+	void               update_timestamps();                                      // this creates local timestamps information vector, update timestamps vector and deviation vector
+	void               init_timeDomains_and_timestamps();                        // this initializes all time domain and timestamps related variables, vectors, and booleans
```
`init_timeDomains_and_timestamps` -> `init_time_domains_and_timestamps`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+	void               init_timeDomains_and_timestamps();                        // this initializes all time domain and timestamps related variables, vectors, and booleans
+	void               get_optimal_time_domain();                                // this gets the optimal time domain which has the minimal value on its max deviation.
+	void               tic(const std::string &input_tag = "Untagged");           // this marks the timestamp begin and partially updates the delta_timestamps
+	void               toc(const std::string &input_tag = "Untagged");           // this marks the timestamp ends and updates the delta_timestamps
```
I like the naming of `tic()` and `toc()`. But maybe something like `timestamp_begin()` and `timestamp_end()` is more descriptive?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+	render_pass_begin_info_0.renderArea.offset.x   = 0;
+	render_pass_begin_info_0.renderArea.offset.y   = 0;
+	render_pass_begin_info_0.clearValueCount       = 2;
+	render_pass_begin_info_0.pClearValues          = clear_values_0.data();
```
`render_pass_begin_info_0` and `clear_values_0` are never used.
Can be removed.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+	render_pass_begin_info_0.clearValueCount       = 2;
+	render_pass_begin_info_0.pClearValues          = clear_values_0.data();
+
+	toc("Creating Command Buffer");
```
This section does not measure the creation of a command buffer but just the initialization of some structs.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+
+	toc("Creating Command Buffer");
+
+	tic("Draw Command Buffer");
```
This section does not measure the drawing of a command buffer, but the recording/building of some command buffers.
You should either change the tags for both measurements, or do just one measurement over the complete `build_command_buffers` function, maybe tagging as "Build Command Buffers"?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+			render_pass_begin_info_1.framebuffer              = offscreen.framebuffer;
+			render_pass_begin_info_1.renderArea.extent.width  = offscreen.width;
+			render_pass_begin_info_1.renderArea.extent.height = offscreen.height;
+			render_pass_begin_info_1.clearValueCount          = 3;
```
Maybe better than plain `3`: `static_cast<uint32_t>(clear_values_1.size())`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+			VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
+			render_pass_begin_info.framebuffer              = filter_pass.framebuffer;
+			render_pass_begin_info.renderPass               = filter_pass.render_pass;
+			render_pass_begin_info.clearValueCount          = 1;
```
Why do you specify two `clear_values`, but use just one?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+			VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
+			render_pass_begin_info.framebuffer              = framebuffers[i];
+			render_pass_begin_info.renderPass               = render_pass;
+			render_pass_begin_info.clearValueCount          = 2;
```
Maybe better than plain `2`, again: `static_cast<uint32_t>(clear_values.size())`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+}
+
+void CalibratedTimestamps::prepare_offscreen_buffer()
+{
```
The current version of the HDR sample uses some algorithm to determine the color format. You should use that here as well.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+		render_pass_create_info.attachmentCount        = static_cast<uint32_t>(attachment_descriptions.size());
+		render_pass_create_info.subpassCount           = 1;
+		render_pass_create_info.pSubpasses             = &subpass;
+		render_pass_create_info.dependencyCount        = 2;
```
Maybe better than plain `2`: `static_cast<uint32_t>(dependencies.size())`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+{
+	{
+		offscreen.width  = static_cast<int>(width);
+		offscreen.height = static_cast<int>(height);
```
You should cast to `int32_t`, not to `int`.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+
+	{
+		filter_pass.width  = static_cast<int>(width);
+		filter_pass.height = static_cast<int>(height);
```
You should cast to `int32_t`, not to `int`.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+	pipeline_create_info.layout       = pipeline_layouts.composition;
+	pipeline_create_info.renderPass   = render_pass;
+	rasterization_state.cullMode      = VK_CULL_MODE_FRONT_BIT;
+	color_blend_state.attachmentCount = 1;
```
`1` -> `static_cast<uint32_t>(blend_attachment_states.size())`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+		update_uniform_buffers();
+}
+
+std::string CalibratedTimestamps::read_time_domain(VkTimeDomainEXT inputTimeDomain)
```
This function does not read anything, but gets you a string. I would call it `to_string`.
Morevover, it's just a local helper function, that does not need to be a member function of `CalibratedTimestamps`.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+	}
+}
+
+void CalibratedTimestamps::update_time_domains()
```
This function just gets you the supported time domains. Maybe just name it `get_time_domains`?
Moreover, are those time domains supposed to vary over time? Why do you call it on every `tic` and `toc`?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+	isTimeDomainUpdated = (result == VK_SUCCESS);
+}
+
+void CalibratedTimestamps::update_timestamps()
```
I would call this function `get_timestamps`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+
+	if (isTimeDomainUpdated && isTimestampUpdated && (time_domain_count > 1))
+	{
+		uint64_t optimal_maxDeviation = *std::min_element(max_deviations.begin(), max_deviations.end());
```
std::min_element gets you the iterator. You don't need to find it again

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,969 @@
+
+		auto iterator_optimal = std::find(max_deviations.begin(), max_deviations.end(), optimal_maxDeviation);
+
+		if (iterator_optimal != max_deviations.end())
```
... and, as long as there's at least one elment, the iterator will never be the `end()`. No need to check that.

asuessenbach reviewed (commented):
Although I'm not a native speaker, I think the readme.md needs some substantial word-smithing.

Besides that, my understanding of the time domains is a bit different than yours. I think, there's nothing like an "optimal" time domain. There is one device time domain and one (Windows) or two (Linux) host device domains. They are not better or worse, they are different domains.
For the device time domain, you need `VkQueueFamilyProperties::timestampValidBits` and `VkPhysicalDeviceLimits::timestampPeriod` to determine nanoseconds.
While the two linux time domains already are in nanoseconds, you need the result from the windows function `QueryPerformanceFrequency` for the windows time domain to get from ticks to seconds.
Finally, the maxDeviation values could be used, somehow, to asses the quality of the actual time stamps.

JeremyG2101 committed:
Based on the comments: 1) Vuklan variables now initialized in proper forms; 2) all name convensions are now unified and up-to-date; 3) name changes for "read_time_domain" to "time_domain_to_string" and is made a local helper function; 4) name changes to update_time_domains() and update_timestamps to get_... 5) get_time_domain will be called only once; 6) tic() and toc() are now renamed as timestamps_begin() and timestamps_end(); 7) get_optimal_time_domain is now renamed to get_device_time_domain and its functionality changed; 8) prepare_offscreen_buffer() is now updated to the latest; 9) all variableCounts are now got using static_cast<uint32_t>variable.size(); 10) single variable usage replacing usage of an array with only one element, and their reference passed, which replaces "array.data()"; 11) syntax changes from UI overlay; 12) now the entire build command buffers will be profilied instead of two saparate parts; 13) static_cast to int for width and height are now fixed by static_cast to int32_t. All code passed the Clang check, however, this is a quick save for code revise only. README.md will be revised per change made.

JeremyG2101 committed:
README.md is revised.

gpx1000 committed:
Merge remote-tracking branch 'downstream/Calibrated_Timestamps_Sample' into vk_Calibrated_Timestamps_Sample

SaschaWillems commented:
This crashes for me at runtime in the `timestamps_begin` function in the following line:

![image](https://user-images.githubusercontent.com/10283127/219451613-5a8f5048-86bc-47d8-876e-3caba4ec4b71.png)

The `timestamps.clear()` there looks wrong, as that vector is then never resized and always has zero elements

Same for `timestamps_end`. The vector is cleared here too and never resized before it's accessed.

SaschaWillems reviewed (changes requested):
See my comment, the sample doesn't work for me at all.

JeremyG2101 commented:
> This crashes for me at runtime in the `timestamps_begin` function in the following line:
> 
> ![image](https://user-images.githubusercontent.com/10283127/219451613-5a8f5048-86bc-47d8-876e-3caba4ec4b71.png)
> 
> The `timestamps.clear()` there looks wrong, as that vector is then never resized and always has zero elements
> 
> Same for `timestamps_end`. The vector is cleared here too and never resized before it's accessed.

Ohh! I forgot to recall that timestamps vector must  be resized and then pass its reference/data to vkGetCalibratedTimestampsEXT.. And for some reason it doesn't crush from my end...
But you are right! Good catch and thanks!

JeremyG2101 committed:
fix a bug in timestamps_begin() and timestamps_end(), where the timestamp vector was cleared without being resized. Its README is revised accordingly.

gpx1000 committed:
Merge remote-tracking branch 'downstream/Calibrated_Timestamps_Sample' into vk_Calibrated_Timestamps_Sample

gpx1000 commented:
This PR is ready for review.

gpx1000 committed:
reword the tutorial.  Rely more upon showing in code than reproducing the code and cleanup the language.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+	if (is_time_domain_updated)
+	{
+		// Get calibrated timestamps:
+		VkResult result = vkGetCalibratedTimestampsEXT(get_device().get_handle(), time_domains.size(), timestamps_info.data(), timestamps.data(), max_deviations.data());
```
Please `static_cast<uint32_t>(time_domains.size())`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+			render_pass_begin_info_1.renderArea.extent.width  = offscreen.width;
+			render_pass_begin_info_1.renderArea.extent.height = offscreen.height;
+			render_pass_begin_info_1.clearValueCount          = static_cast<uint32_t>(clear_values_1.size());
+			render_pass_begin_info_1.pClearValues             = clear_values_1.data();
```
Maybe move clear_values_1 and render_pass_begin_info_1 out of the loop.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		{
+			VkClearValue clear_value{};
+			clear_value.color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+			clear_value.depthStencil = {0.0f, 0};
```
`VkClearValue` is a union; setting `depthStencil` overwrites the `color`.
You should probably stay with the `color`.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		VkSubpassDescription subpass    = {};
+		subpass.pipelineBindPoint       = VK_PIPELINE_BIND_POINT_GRAPHICS;
+		subpass.pColorAttachments       = color_references.data();
+		subpass.colorAttachmentCount    = 2;
```
`2` -> `static_cast<uint32_t>(color_references.size())`

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		VkSubpassDescription subpass = {};
+		subpass.pipelineBindPoint    = VK_PIPELINE_BIND_POINT_GRAPHICS;
+		subpass.pColorAttachments    = color_references.data();
+		subpass.colorAttachmentCount = 1;
```
`1` -> `static_cast<uint32_t>(color_references.size())`

asuessenbach replied:
You had marked this as resolved without any change. Do you want to keep it like this? If so, why?

gpx1000 replied:
that's my fault; I could have sworn I got it but I definitely missed it.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+	if (result == VK_SUCCESS)
+	{
+		// Initialize time domain vector
+		time_domains.clear();
```
No need to clear `time_domains`, as this is called just once.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		// Initialize time domain vector
+		time_domains.clear();
+		// Resize time domains vector:
+		time_domains.resize(static_cast<int>(time_domain_count));        // this needs static cast to int
```
Why should you need to cast `time_domain_count` to `int`?

asuessenbach replied:
You should as well remove the comment.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		// Resize time stamps vector
+		timestamps.resize(static_cast<int>(time_domain_count));
+		// Resize max deviations vector
+		max_deviations.resize(static_cast<int>(time_domain_count));
```
No need to `static_cast<int>` here

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+	}
+
+	// Time domain is successfully updated:
+	is_time_domain_updated = ((result == VK_SUCCESS) && (time_domain_count > 0));
```
`is_time_domain_updated` sounds like if it is, well, updated, regularly, after some amount of time.
But it is just that everything is initialized/prepared once to actually get the data.

asuessenbach replied:
You should as well adjust the comment.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		VkResult result = vkGetCalibratedTimestampsEXT(get_device().get_handle(), time_domains.size(), timestamps_info.data(), timestamps.data(), max_deviations.data());
+
+		// Timestamps is successfully updated:
+		is_timestamp_updated = (result == VK_SUCCESS);
```
Maybe change this function to return a bool `(result == VK_SUCCESS)`.
No need to have that `is_timestamp_updated` anymore.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+	}
+}
+
+void CalibratedTimestamps::init_time_domains_and_timestamps()
```
Is this function useful at all?
It's just some one-time member variable initialization.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		// Create a local delta_timestamp to push back to the vector delta_timestamps
+		DeltaTimestamp delta_timestamp{};
+
+		// Naming the tic-toc tags and begin timestamp for this particular mark
```
tic-toc ?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -0,0 +1,988 @@
+		}
+		delta_timestamp.begin = timestamps[selected_time_domain.index];
+
+		// Push back this partially filled element to the vector, which will be filled in its corresponding toc
```
toc ?

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+		uint64_t    delta = 0;
+		std::string tag   = "Untagged";
+
+		void get_delta()
```
Is this really worth a function?
If at all, it doesn't get you anything (returning void) but calculates something. So `calculate_delta` or `determine_delta` would be a better fit.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.h:
```diff
@@ -0,0 +1,190 @@
+	std::vector<uint64_t>        timestamps{};                          // timestamps vector
+	std::vector<uint64_t>        max_deviations{};                      // max deviations vector
+
+	struct
```
Some structs are named, others are not.
Is there a rationale behind that?

asuessenbach reviewed (commented):
This sample just shows how to use the CalibratedTimestamps functions. If that was your goal, I'm fine with it.
But I see three possible improvements:
- With the `std::vector<DeltaTimestamp> delta_timestamps`, you are prepared to take multiple timestamps per frame. But you don't do it. Maybe you should consider adding one or more additional timestamps?
- You explicitly select the device domain, but measure the recording (or building) of a command buffer, which probably happens completely on the host. Maybe you should measure some device work, instead, like drawing or buffer uploading?
- You don't show any use of the deviations. Maybe at least displaying it could get you some feeling on what they are?

gpx1000 committed:
address most of the comments.

gpx1000 commented:
I tend to agree this sample doesn't fully exercise calibrated timestamps.  However, it does a fairly good job of demonstrating how to use them which I think is a decent goal.  I'll have Jeremy do another pass at increasing the scope a little bit more.  My only concern is about ensuring this sample's core is to teach specifically how to do calibrated timestamps and not get too complex to ultimately be useful.  Thanks for the review!

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -373,7 +373,7 @@ void CalibratedTimestamps::prepare_offscreen_buffer()
 		render_pass_create_info.pAttachments           = attachment_descriptions.data();
 		render_pass_create_info.attachmentCount        = static_cast<uint32_t>(attachment_descriptions.size());
-		render_pass_create_info.subpassCount           = 1;
+		render_pass_create_info.subpassCount           = static_cast<uint32_t>(color_references.size());
```
the subpassCount is not related to the number of color_references! Keep the `1` here.

asuessenbach commented on samples/extensions/calibrated_timestamps/calibrated_timestamps.cpp:
```diff
@@ -109,24 +110,24 @@ void CalibratedTimestamps::build_command_buffers()
 	render_pass_begin_info_0.clearValueCount       = static_cast<uint32_t>(clear_values_0.size());
 	render_pass_begin_info_0.pClearValues          = clear_values_0.data();
 
```
One could go even further here, and move all the `VkClearValue`s, `VkRenderPassBeginInfo`s, `VkViewport`s, and `VkRect2D`s out of the loop, renaming them for example to `offscreen*`, `bloom*` or `filter_pass*`, and `framebuffer*`.
Would make the actual recording loop very short and clear.

gpx1000 committed:
Sorry about that could have sworn I did this but I missed it.

gpx1000 committed:
move intransitive variables out of the for loop.

gpx1000 committed:
fix the comments.

gpx1000 committed:
Merge branch 'main' into vk_Calibrated_Timestamps_Sample

gpx1000 committed:
show using multiple timestamp captures.

We probably should move away from using a vector and to a map or something and use hash for indexes.  However, this will work for such a small sample.

gpx1000 committed:
clang format

gpx1000 committed:
switch to using an unordered map and add more samplers.

gpx1000 committed:
remove errant usleep.

gary-sweet reviewed (approved):
I just added calibrated timestamp support to my driver, so was actually able to test this.

marty-johnson59 commented:
Merging w/ 3 approvals
